frame/support/src/traits/hooks.rs

```diff
@@ -80,13 +80,24 @@ pub trait OnIdle<BlockNumber> {
 }
 
 #[impl_for_tuples(30)]
-impl<BlockNumber: Clone> OnIdle<BlockNumber> for Tuple {
+impl<BlockNumber: Clone + AtLeast32BitUnsigned> OnIdle<BlockNumber> for Tuple {
 	fn on_idle(n: BlockNumber, remaining_weight: crate::weights::Weight) -> crate::weights::Weight {
-		let mut weight = 0;
+		let mut on_idle_functions = sp_std::vec::Vec::<
+			fn(BlockNumber, crate::weights::Weight) -> crate::weights::Weight,
+		>::new();
 		for_tuples!( #(
-			let adjusted_remaining_weight = remaining_weight.saturating_sub(weight);
-			weight = weight.saturating_add(Tuple::on_idle(n.clone(),  adjusted_remaining_weight));
+			on_idle_functions.push(Tuple::on_idle);
 		)* );
+		let mut weight = 0;
+		for pallet_index in 0..on_idle_functions.len() {
+			let adjusted_remaining_weight = remaining_weight.saturating_sub(weight);
+			let n1 = n.clone() % BlockNumber::from(on_idle_functions.len() as u32);
+			let n1 : u32 = n1.try_into().unwrap_or_else(|_| panic!("cannot convert BlockNumber to u32"));
+			let index =
+				((pallet_index as u32) + n1) % (on_idle_functions.len() as u32);
```

Comment:
This can all be moved up.

`n1.try_into().unwrap_or_else(|_| panic!("cannot convert BlockNumber to u32"));` this can be made into a proof by converting it into an `usize`. `on_idle_functions.len()` can never be more than `usize`.

And then you do the following loop:

`for on_idle in on_idle_functions.iter().cycle().skip(start_index).take(on_idle_functions.len())`